doc/man7/EVP_KDF-ARGON2.pod

```diff
@@ -0,0 +1,150 @@
+=pod
+
+=head1 NAME
+
+EVP_KDF-ARGON2 - The Argon2 EVP KDF implementation
+
+=head1 DESCRIPTION
+
+Support for computing the B<argon2> password-based KDF through the B<EVP_KDF>
+API.
+
+The EVP_KDF-ARGON2 algorithm implements the Argon2 password-based key
+derivation function, as described in IETF RFC 9106.  It is memory-hard in
+the sense that it deliberately requires a significant amount of RAM for efficient
+computation. The intention of this is to render brute forcing of passwords on
+systems that lack large amounts of main memory (such as GPUs or ASICs)
+computationally infeasible.
+
+=head2 Supported parameters
+
+The supported parameters are:
+
+=over 4
+
+=item "pass" (B<OSSL_KDF_PARAM_PASSWORD>) <octet string>
+
+=item "salt" (B<OSSL_KDF_PARAM_SALT>) <octet string>
+
+=item "secret" (B<OSSL_KDF_PARAM_SECRET>) <octet string>
+
+=item "iter" (B<OSSL_KDF_PARAM_ITER>) <unsigned integer>
+
+These parameters work as described in L<EVP_KDF(3)/PARAMETERS>.
+
+=item "t" (B<OSSL_KDF_PARAM_THREADS>) <unsigned integer>
+
+The number of threads. This can only be used with built-in thread support.
+Threading must be explicitly enabled. See EXAMPLES section for more information.
+
+=item "s" (B<OSSL_KDF_PARAM_ARGON2_SZ>) <unsigned integer>
+
+Digest length.
+
+=item "a" (B<OSSL_KDF_PARAM_ARGON2_AD>) <octet string>
+
+Optional associated data.
+
+=item "l" (B<OSSL_KDF_PARAM_ARGON2_LANES>) <unsigned integer>
+
+Number of lanes.
+
+=item "m" (B<OSSL_KDF_PARAM_ARGON2_MEMCOST>) <unsigned integer>
+
+Memory cost parameter (the number of 1k memory blocks used).
+
+=item "v" (B<OSSL_KDF_PARAM_ARGON2_VERSION>) <unsigned integer>
+
+Argon2 version. Supported values: 0x10, 0x13 (default).
+
+=item "f1" (B<OSSL_KDF_PARAM_ARGON2_FLAG_CLEAR_PWD>) <unsigned integer>
+
+If set (nonzero), B<OPENSSL_cleanse> is called on password during initial
+hash phase.
+
+=item "f2" (B<OSSL_KDF_PARAM_ARGON2_FLAG_CLEAR_SECRET>) <unsigned integer>
+
+If set (nonzero), B<OPENSSL_cleanse> is called on secret during initial
+hash phase.
```

Comment:
These flags have different names in the core_names.h file. Also I'm not sure we want them. What is the rationale? No other KDF implementation has anything like this.

Reply:
The rationale for these flags is:

These inputs are used to calculate the initial hash H_0 during `kdf_argon2_derive` in `initial_hash` and they are not required afterwards for the current derivation. They get ultimately zeroed with `OPENSSL_clear_free` in `kdf_argon2_free` on free, `kdf_argon2_reset` on reset, or `kdf_argon2_ctx_set_pwd` on that param change.

If one wishes to derive a new key with, for example, a different ad, one may just provide a new octet string param `OSSL_KDF_PARAM_ARGON2_AD` and call derive again. This saves the user from storing the password or secret after having initialized the Argon2 context.

If that's not required one only has to explicitly set these clear flags and `OPENSSL_cleanse` is called right after initialization, not after any of the above operations which may occur substantially later.

Reply:
Could you please put some of the rationale into the manual page text?  Something like: `This can be useful if multiple keys with different ad value are to be generated from a single password and secret.`

Reply:
Certainly, would something like this work?
```
=item "f1" (B<OSSL_KDF_PARAM_ARGON2_FLAG_CLEAR_PWD>) <unsigned integer>

=item "f2" (B<OSSL_KDF_PARAM_ARGON2_FLAG_CLEAR_SECRET>) <unsigned integer>

If set (non-zero), password (or secret) stored in Argon2 context is zeroed
early during initial hash computation, as soon as it's not needed.

It's zeroed along the rest of Argon2 context data on clear, free, reset,
password (or secret) param assignment operations otherwise.

This can be useful if, for example, multiple keys with different ad value
are to be generated from a single password and secret.
```

Reply:
Hmm, I am still not convinced it is that useful because if I am not going to reuse the KDF CTX to derive multiple keys I'll probably call EVP_KDF_CTX_free() just after the derive call and thus the stored password/secret is zeroized immediately after the derivation anyway. But if I want to reuse the context to derive multiple keys later I cannot use these flags. Or do I misunderstand?

Reply:
You're right, if you want to re-use the context, you have to do it before a call to EVP_KDF_CTX_free is made and you cannot use these flags.

If low memory cost/iter count is used, derive will likely finish quickly, however, derive can take quite a while otherwise (especially with large memcost). I think that's when you might want to use these, if you don't plan on genering multiple keys, no point for the data to hang around.

Reply:
(I suppose we could merge them into a single flag that could be re-used elsewhere, if adding too many flags is a concern.)

Reply:
Refactored the two tags into a generic `OSSL_KDF_PARAM_EARLY_CLEAN`.